pkg/analyze/data_source_test.go

```diff
@@ -0,0 +1,265 @@
+				t.Fatalf("expected '%v' got: '%v'", reflect.TypeFor[dummyLister](), reflect.TypeOf(lister))
+			}
+
+			actualObjects, err := lister.List()
+			if err != nil {
+				t.Fatalf("dummyLister.List returned an error: %v", err)
+			}
+
+			assertExpectedContainsPermutatedActual(t, test.Objects, actualObjects)
+		})
+	}
+}
+
+func compareSingle[T any, U *T](
+	t *testing.T,
+	all []runtime.Object,
+	listFunc func(selector labels.Selector) (ret []*T, err error),
+) {
+	actualResources, err := listFunc(labels.Everything())
+	if err != nil {
+		t.Fatalf("%s lister returned an error: %v", reflect.TypeFor[U]().Name(), err)
+	}
+
+	var expectedResources []U
+	for _, res := range all {
+		switch v := res.(type) {
+		case U:
+			expectedResources = append(expectedResources, v)
+		}
+	}
+
+	assertExpectedContainsPermutatedActual(t, expectedResources, actualResources)
+}
+
+func TestNewDataSourceFromClients(t *testing.T) {
+	newDataSourceTests := map[string]struct {
+		kubernetesObjects []runtime.Object
+		scyllaObjects     []runtime.Object
+	}{
+		"no objects": {},
+		"some listers empty": {
+			kubernetesObjects: []runtime.Object{
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod1",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod2",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Svc1",
+						Namespace: "NS1",
+					},
+				},
+			},
+			scyllaObjects: []runtime.Object{
+				&scyllav1.ScyllaCluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "ScyllaCluster1",
+						Namespace: "NS1",
+					},
+				},
+			},
+		},
+		"all objects": {
+			kubernetesObjects: []runtime.Object{
+				&v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Pod1",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Svc1",
+						Namespace: "NS1",
+					},
+				},
+				&v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "Secret1",
+						Namespace: "NS2",
+					},
+				},
+				&v1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "ConfigMap1",
+						Namespace: "NS1",
+					},
+				},
+			},
+			scyllaObjects: []runtime.Object{
+				&scyllav1.ScyllaCluster{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "ScyllaCluster1",
+						Namespace: "NS1",
+					},
+				},
+			},
+		},
+	}
+
+	for name, test := range newDataSourceTests {
+		t.Run(name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			fakeClient := kubefake.NewSimpleClientset(test.kubernetesObjects...)
+			scyllaFakeClient := fake.NewSimpleClientset(test.scyllaObjects...)
+			defer cancel()
+
+			dataSource, err := NewDataSourceFromClients(ctx, fakeClient, scyllaFakeClient)
+			if err != nil {
+				t.Fatalf("NewDataSourceFromClients returned an error: %v", err)
+			}
```

Comment:
the more idiomatic approach would be to add an expected err to test struct and compare expected err vs actual err instead, since it's returned by the function being tested

Reply:
done